{\itshape align2bed} is a light-\/weight tool that extracts single nucleotide polymorphisms (S\+N\+Ps) from F\+A\+S\+TA alignments and saves them to a \href{http://zzz.bwh.harvard.edu/plink/index.shtml}{\tt plink} binary-\/format B\+ED file. It is fast and can process whole-\/genome data on a laptop in minutes. The F\+A\+S\+TA alignments should be in the format provided by the {\itshape Drosophila} Genome Nexus (\href{http://www.johnpool.net/genomes.html}{\tt http\+://www.\+johnpool.\+net/genomes.\+html}), i.\+e. the data for each line (or individual) are in a separate file with nucleotides in one line and without the customary F\+A\+S\+TA header. Each chromosome arm has its own set of files.

The sofware uses control files that list paths to each F\+A\+S\+TA file to be processed. An example data set is included to illustrate the necessary features of the data. One of the individuals must be marked as the reference, and it is also assumed that this reference is the outgroup. The reference genotypes are not included in the output. B\+ED format files require the S\+N\+Ps to be biallelic, so S\+N\+Ps that do not meet this criterion are not included as output. However, if a S\+NP is biallelic within the population (non-\/reference) sample, but both alleles are different from the outgroup, it is included. Names of such S\+N\+Ps are marked with \char`\"{}d\char`\"{} at the end (in the accompanying .bim file) to enable downstream filtering. In addition, S\+N\+Ps with outgroup missing are included but their names marked with \char`\"{}m\char`\"{} at the end. S\+NP names are \char`\"{}s\char`\"{} plus position, then \char`\"{}m\char`\"{} or \char`\"{}d\char`\"{} if applicable, then underscore (\char`\"{}\+\_\+\char`\"{}), then chromosome arm name.

While {\itshape align2bed} is tailored for the {\itshape Drosophila} Genome Nexus data, there are three ways it can be extended to similar data sets from other species. Data can be arranged to mimic the {\itshape Drosophila} set by treating chromosomes in groups of five. Alignment length can vary indefinitely. Furthermore, I wrote the program using a class that has wider applicability. Taking the {\itshape align2bed} source code as an exmaple, and reading the provided interface documentation, someone with even very limited experience in C++ can write software that applies to different data sets and hardware configurations. Finally, anyone who would like to extend functionality even further is welcome to modify the class implementation to suit their needs.

Neither {\itshape align2bed} itself nor the class used to implement it has any dependencies outside of the C++ S\+TL. Only a compiler capable of recognizing the C++11 standard is required. The implementation is multithreaded, with chromosome arms processed in parallel. Each thread allocates a 2\+Gb buffer to read the F\+A\+S\+TA files. The {\itshape align2bed} source can be easily modified to change the threading and memory allocation parameters (see included class documentation for details).

To compile, make sure you are in the directory with the source code files and run \begin{DoxyVerb}g++ align2bed.cpp sequence.cpp -o align2bed -O3 -march=native -std=c++11
\end{DoxyVerb}


then copy the binary where you need it. Run by typing {\ttfamily ./align2bed} in the directory with the binary and the data, or move into an appropriate /bin folder for global access. 